### PR TITLE
Create LDAP identity service

### PR DESCRIPTION
**Pull request by cychiuae**
ref DEV-1652

### CODE REVIEW

File: migrations/authgear/20240731135615-add_ldap_identity.sql [hunk 1]
```diff
@@ -0,0 +1,17 @@
+-- +migrate Up
+CREATE TABLE _auth_identity_ldap
+(
+    id                 text  PRIMARY KEY REFERENCES _auth_identity (id),
+    app_id             text  NOT NULL,
+    server_name        text  NOT NULL,
+    user_id_attribute  text  NOT NULL,
```
Comment (cychiuae) [inline migrations/authgear/20240731135615-add_ldap_identity.sql:7]: Change to `user_id_attribute_oid`

File: pkg/lib/authn/identity/ldap_identity.go [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+	"github.com/authgear/authgear-server/pkg/api/model"
+)
+
+type LDAP struct {
+	ID                   string                 `json:"id"`
+	CreatedAt            time.Time              `json:"created_at"`
+	UpdatedAt            time.Time              `json:"updated_at"`
+	UserID               string                 `json:"user_id"`
+	ServerName           string                 `json:"server_name"`
+	UserIDAttribute      string                 `json:"user_id_attribute"`
```
Comment (cychiuae) [inline pkg/lib/authn/identity/ldap_identity.go:15]: Change to `user_id_attribute_oid`

File: pkg/lib/authn/identity/ldap_spec.go [hunk 3]
```diff
@@ -0,0 +1,9 @@
+package identity
+
+type LDAPSpec struct {
+	ServerName           string                 `json:"server_name"`
+	UserIDAttribute      string                 `json:"user_id_attribute"`
```
Comment (cychiuae) [inline pkg/lib/authn/identity/ldap_spec.go:5]: Change to `user_id_attribute_oid`

File: pkg/lib/authn/identity/ldap/store.go [hunk 4]
```diff
@@ -0,0 +1,314 @@
...
+	}
+
+	return is, nil
+
+}
+
+func (s *Store) GetByServerUserID(serverName string, userIDAttributeOID string, userIDAttributeValue string) (*identity.LDAP, error) {
+	q := s.selectQuery().
+		Where(
+			"o.servier_name = ? AND o.user_id_attribute_oid = ? AND o.user_id_attribute_value = ?",
```
Comment (cychiuae) [inline pkg/lib/authn/identity/ldap/store.go:178]: typo `servier_name` -> `server_name`

File: pkg/lib/authn/identity/service/service.go [hunk 5]
```diff
@@ -456,6 +516,17 @@ func (s *Service) ListByUserIDs(userIDs []string) (map[string][]*identity.Info,
 		}
 	}
 
+	// ldap
+	if ldapRefs, ok := refsByType[model.IdentityTypeLDAP]; ok && len(ldapRefs) > 0 {
+		ldapIdens, err := s.SIWE.GetMany(extractIDs(ldapRefs))
```
Comment (cychiuae) [inline pkg/lib/authn/identity/service/service.go:521]: `s.SIWE` -> `s.LDAP`

File: pkg/lib/authn/identity/ldap/store.go [hunk 6]
```diff
@@ -125,6 +126,51 @@ func (s *Store) List(userID string) ([]*identity.LDAP, error) {
...
+		if err != nil {
+			return nil, err
+		}
+		is = append(is, i)
+	}
+
+	return is, nil
+}
+
+func (s *Store) ListByClaimJSONPointer(pointer jsonpointer.T, value string) ([]*identity.LDAP, error) {
```
Comment (tung2744) [inline pkg/lib/authn/identity/ldap/store.go:152]: Please remove this method.
You can see https://github.com/authgear/authgear-server/pull/4531/files

Now we always use ListByClaim

Comment (cychiuae) [reply pkg/lib/authn/identity/ldap/store.go:152]: I know. I am waiting louis help me rebasing the branch `feature-ldap`.

File: migrations/authgear/20240731135615-add_ldap_identity.sql (hunk 1 above)
Comment (louischan-oursky) [inline migrations/authgear/20240731135615-add_ldap_identity.sql]: Change to `user_id_attribute_name`

File: migrations/authgear/20240731135615-add_ldap_identity.sql [hunk 7]
```diff
@@ -0,0 +1,17 @@
...
+    user_id_value      text  NOT NULL,
```
Comment (louischan-oursky) [inline migrations/authgear/20240731135615-add_ldap_identity.sql]: Change to `user_id_attribute_value`

File: migrations/authgear/20240731135615-add_ldap_identity.sql [hunk 8]
```diff
@@ -0,0 +1,17 @@
...
+    claims             jsonb NOT NULL,
+    raw_entry_json     jsonb NOT NULL
+);
+ALTER TABLE _auth_identity_ldap
+    ADD CONSTRAINT _auth_identity_ldap_unique UNIQUE (app_id, server_name, user_id_attribute, user_id_value);
```
Comment (louischan-oursky) [inline migrations/authgear/20240731135615-add_ldap_identity.sql]: Add 3 more indice

```
CREATE INDEX _auth_identity_ldap_claim_preferred_username ON _auth_identity_ldap (app_id, (claims ->> 'preferred_username'));
CREATE INDEX _auth_identity_ldap_claim_phone_number ON _auth_identity_ldap (app_id, (claims ->> 'phone_number'));
CREATE INDEX _auth_identity_ldap_claim_email ON _auth_identity_ldap (app_id, (claims ->> 'email'));
```

File: pkg/lib/authn/identity/ldap_identity.go (hunk 2 above)
Comment (louischan-oursky) [reply pkg/lib/authn/identity/ldap_identity.go]: Change to `user_id_attribute_name`

File: pkg/lib/authn/identity/ldap_spec.go (hunk 3 above)
Comment (louischan-oursky) [reply pkg/lib/authn/identity/ldap_spec.go]: Change to `user_id_attribute_name`

File: pkg/lib/authn/identity/service/service.go [hunk 9]
```diff
@@ -89,6 +89,24 @@ type SIWEIdentityProvider interface {
 	Delete(i *identity.SIWE) error
 }
 
+type LDAPIdentityProvider interface {
+	Get(userID, id string) (*identity.LDAP, error)
+	GetMany(ids []string) ([]*identity.LDAP, error)
+	List(userID string) ([]*identity.LDAP, error)
+	GetByServerUserID(serverName string, userIDAttribute string, userIDAttributeValue string) (*identity.LDAP, error)
```
Comment (louischan-oursky) [inline pkg/lib/authn/identity/service/service.go]: Rename `userIDAttribute` to `userIDAttributeName`

File: pkg/lib/authn/identity/ldap/store.go [hunk 10]
```diff
@@ -97,3 +97,24 @@ func (s *Store) GetMany(ids []string) ([]*identity.LDAP, error) {
 
 	return is, nil
 }
+
+func (s *Store) List(userID string) ([]*identity.LDAP, error) {
+	builder := s.selectQuery().Where("p.user_id = ANY (?)", pq.Array(userID))
```
Comment (louischan-oursky) [inline pkg/lib/authn/identity/ldap/store.go:109]: Prefer `p.user_id = ?`

File: pkg/lib/authn/identity/ldap/store.go [hunk 11]
```diff
@@ -118,3 +118,18 @@ func (s *Store) List(userID string) ([]*identity.LDAP, error) {
 
 	return is, nil
 }
+
+func (s *Store) GetByServerUserID(serverName string, userIDAttribute string, userIDAttributeValue string) (*identity.LDAP, error) {
+	q := s.selectQuery().
+		Where(
+			"o.server_name = ? AND o.user_id_attribute = ? AND o.user_id_attribute_value = ?",
```
Comment (louischan-oursky) [inline pkg/lib/authn/identity/ldap/store.go]: The alias you use is `l`, not `o`.

File: pkg/lib/authn/identity/ldap/provider.go [hunk 12]
```diff
@@ -0,0 +1,74 @@
...
+}
+
+func (p *Provider) GetByServerUserID(serverName string, userIDAttribute string, userIDAttributeValue string) (*identity.LDAP, error) {
+	return p.Store.GetByServerUserID(serverName, userIDAttribute, userIDAttributeValue)
+}
+
+func (p *Provider) New(
+	userID string,
+	serverName string,
+	userIDAttribute string,
```
[the remark this answers is not in the document]
Comment (louischan-oursky) [reply pkg/lib/authn/identity/ldap/provider.go]: Change to `userIDAttributeName`

File: pkg/lib/authn/identity/ldap/provider.go [hunk 13]
```diff
@@ -25,3 +26,22 @@ func (p *Provider) List(userID string) ([]*identity.LDAP, error) {
...
+	rawEntryJSON map[string]interface{},
+) *identity.LDAP {
+	return &identity.LDAP{
+		ID:                   uuid.New(),
+		UserID:               userID,
+		ServerName:           serverName,
+		UserIDAttribute:      userIDAttribute,
+		UserIDAttributeValue: userIDAttributeValue,
+		Claims:               claims,
+		RawEntryJSON:         rawEntryJSON,
```
Comment (louischan-oursky) [inline pkg/lib/authn/identity/ldap/provider.go:52]: Make sure we do not save NULL to  these two JSONB column. I think the best place to do this is in this method.

```
if claims == nil {
  claims = make(map[string]interface{})
}
if rawEntryJSON == nil {
  rawEntryJSON = make(map[string]interface{})
}
```

File: pkg/lib/authn/identity/claim_key.go [hunk 14]
```diff
@@ -42,8 +42,8 @@ const (
 
 	// IdentityClaimLDAPServerName is a claim with a string value.
 	IdentityClaimLDAPServerName string = "https://authgear.com/claims/ldap/server_name"
-	// IdentityClaimLDAPUserIDAttributeName is a claim with a string value.
-	IdentityClaimLDAPUserIDAttributeName string = "https://authgear.com/claims/ldap/user_id_attribute_name"
+	// IdentityClaimLDAPUserIDAttributeOID is a claim with a string value.
+	IdentityClaimLDAPUserIDAttributeOID string = "https://authgear.com/claims/ldap/user_id_attribute_oid"
```
Comment (louischan-oursky) [inline pkg/lib/authn/identity/claim_key.go:46]: Rename it back to `user_id_attribute_name`

File: pkg/lib/authn/identity/ldap/store.go [hunk 15]
```diff
@@ -39,7 +39,8 @@ func (s *Store) selectQuery() db.SelectBuilder {
 			"l.raw_entry_json",
 		).
 		From(s.SQLBuilder.TableName("_auth_identity"), "p").
-		Join(s.SQLBuilder.TableName(tableNameAuthIdentityLDAP), "l", "p.id = l.id")
+		Join(s.SQLBuilder.TableName(tableNameAuthIdentityLDAP), "l", "p.id = l.id").
+		OrderBy("p.created_at ASC")
```
Comment (louischan-oursky) [inline pkg/lib/authn/identity/ldap/store.go:43]: Do not add this now, to stay consistent with other stores.

File: pkg/lib/authn/identity/ldap_identity.go [hunk 16]
```diff
@@ -41,8 +41,17 @@ func (i *LDAP) ToLDAPSpec() *LDAPSpec {
...
+	if email, ok := i.Claims[string(model.ClaimEmail)].(string); ok {
+		claims[model.ClaimEmail] = email
+	}
+	if phoneNumber, ok := i.Claims[string(model.ClaimPhoneNumber)].(string); ok {
+		claims[model.ClaimPhoneNumber] = phoneNumber
+	}
+	if username, ok := i.Claims[string(model.ClaimPreferredUsername)].(string); ok {
+		claims[model.ClaimPreferredUsername] = username
+	}
+	return claims
```
Comment (louischan-oursky) [inline pkg/lib/authn/identity/ldap_identity.go:56]: OAuth identity should follow this logic. Please open another PR for this.

**Review by louischan-oursky — CHANGES_REQUESTED** [review]

File: pkg/lib/authn/identity/ldap_identity.go (hunk 16 above)
Comment (cychiuae) [reply pkg/lib/authn/identity/ldap_identity.go:56]: Handled in #4538

**Comment by cychiuae** [comment]
@louischan-oursky updated thanks

**Review by louischan-oursky — APPROVED** [review]

**Pull request merged**